### PR TITLE
6X: Fix tcp interconnect hang

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
There was a hang like this: when one QE errors out before 'SetupInterconnect',
QD will keep waiting for the incoming connections to be established and doesn't
check the error message from dispatcher. Other QEs are finished and hang in
function 'waitOnOutbound'.

Co-authored-by: Asim R P <apraveen@pivotal.io>
Co-authored-by: Gang Xiong <gxiong@pivotal.io>
Co-authored-by: Ning Yu <nyu@pivotal.io>

## Background
This is a re-push of https://github.com/greenplum-db/gpdb/pull/8277 and https://github.com/greenplum-db/gpdb/pull/8588, they were reverted as they broke the 6X pipelines because an `assert()` check added by #8277 caught some incomplete initializations. 

The first commit in current PR is a combination of those two PRs; the second commit in current PR is to fix the initialization issues, tests are not provided as the issues are covered by existing ICW tests.

A devel pipeline is setup for verification.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/executor/execMain.c [hunk 1]
```diff
@@ -262,6 +262,43 @@ ExecutorStart(QueryDesc *queryDesc, int eflags)
...
+ * Check whether a plan is parallel by counting motions.
+ *
+ * A plan itself and its init plans can all contain motions, when itself
+ * contains at least one motion then it is a parallel plan.  In this function
+ * we count the motions of the plan itself to tell whether it is a parallel
+ * plan.
+ *
+ * NOTE: plan->dispatch is not checked by this function.
+ */
+static bool
```
Comment (kainwen) [inline src/backend/executor/execMain.c]: Do we need to add declaration at the head of this file?  I do not find it in the diff.

Comment (pivotal-ning-yu) [reply src/backend/executor/execMain.c]: Good point, I'll add it.

**Comment by pivotal-ning-yu** [comment]
Force pushed the patches, added the declaration of the new function.

**Comment by pivotal-ning-yu** [comment]
Thanks for the review, merge it now.

**Comment by hlinnaka** [comment]
Hi! This conflicted with my https://github.com/greenplum-db/gpdb/pull/8612, so I had to look into this. I find the `planIsParallel()` pretty weird. The executor doesn't seem like the right place for that. The point of the existing 'dispatch' and 'nMotionNodes' fields is to tell the executor about these things, so if they're not currently carrying the information the executor needs, then let's fix that.

I'm confused by the commit message:

    Check parallel plans correctly
    
    In standard_ExecutorStart() we should dispatch a plan if it is parallel,
    currently this is determined by checking planTree->dispatch is
    DISPATCH_PARALLEL or not.  However sometimes a DISPATCH_UNDETERMINED
    plan can also be parallel.  For example:
    
            CREATE TABLE arrtest_f (f0 int, f1 text, f2 float8)
              DISTRIBUTED RANDOMLY;
            EXPLAIN
            SELECT ARRAY(select f2 from arrtest_f order by f2) AS "ARRAY"
              ORDER BY 1;
                                 QUERY PLAN
            --------------------------------------------------
             Result
               InitPlan 1 (returns $0)  (slice2)
                 ->  Gather Motion 3:1  (slice1; segments: 3)
                       Merge Key: f2
                       ->  Sort
                             Sort Key: f2
                             ->  Seq Scan on arrtest_f
             Optimizer: Postgres query optimizer
            (8 rows)
    
    To fix it we should also check whether the plan contains motions.  Note
    that we should only check for the motions of itself, the motions of its
    init plans should not be counted.
    
The main Plan above is in fact *not* parallel, DISPATCH_UNDETERMINED seems correct. The InitPlan does contain Motions, and should indeed be run in parallel. I think the real culprit is in the second check that was modified by this PR:
```
@@ -705,7 +741,9 @@ standard_ExecutorStart(QueryDesc *queryDesc, int eflags)
                         *
                         * Main plan is parallel, send plan to it.
                         */
-                       if (queryDesc->plannedstmt->planTree->dispatch == DISPATCH_PARALLEL)
+                       if (queryDesc->plannedstmt->planTree->dispatch == DISPATCH_PARALLEL ||
+                               planIsParallel(queryDesc->plannedstmt,
+                                                          queryDesc->plannedstmt->planTree))
                                CdbDispatchPlan(queryDesc, needDtxTwoPhase, true);
                }
 
@@ -738,7 +776,9 @@ standard_ExecutorStart(QueryDesc *queryDesc, int eflags)
                {
                        /* Run a root slice. */
                        if (queryDesc->planstate != NULL &&
-                               queryDesc->planstate->plan->nMotionNodes > 0 && !estate->es_interconnect_is_setup)
+                               planIsParallel(queryDesc->plannedstmt,
+                                                          queryDesc->planstate->plan) &&
+                               !estate->es_interconnect_is_setup)
                        {
                                Assert(!estate->interconnect_context);
                                SetupInterconnect(estate);
```
Notice how the first check checks if `dispatch == DISPATCH_PARALLEL`, but second one does not. I think *that* was wrong. We should not set up the interconnect, even if there are Motions, if the plan is not marked with DISPATCH_PARALLEL. That implies that the Motions are in the Init Plans, like you also check in planIsParallel(), but checking for DISPATCH_PARALLEL is much more straightforward.

I tested reverting commit 7d74aa5527, and run that test case from the commit message. This happens:
```
FATAL:  Unexpected internal error (cdbdisp_async.c:138)
DETAIL:  FailedAssertion("!(ds)", File: "cdbdisp_async.c", Line: 138)
HINT:  Process 14724 will wait for gp_debug_linger=120 seconds before termination.
```
Adding the `dispatch == DISPATCH_PARALLEL` check to that second hunk fixes that problem, just as well as this PR.

I'm going to create a new PR to revert commit 7d74aa5527, and do that simpler fix. https://github.com/greenplum-db/gpdb/pull/8612 is going to remove `Plan->nMotions` field, so that needs to go in any case.

**Comment by hlinnaka** [comment]
Question: That query mentioned in the commit message is from the 'arrays' regression test. Why did it suddenly start failing?

**Comment by pivotal-ning-yu** [comment]
> Question: That query mentioned in the commit message is from the 'arrays' regression test. Why did it suddenly start failing?

Thank you for your analysis and the better fix.  The `arrays` test, as well as many other tests, didn't fail because the `Assert(ds)` check was added by the first commit in this PR, the seconds commit is just to fix the failures.  So I believe the interconnect setup condition has been wrong for a long time, but that didn't result in actual runtime errors until now.

I'll look into your PRs, please give me some time to review them.

**Pull request merged**